File: resources/logdb/MesgShred.py
import re
import sys
from os import path
import src.python.logdb.createdb as createdb
from src.python.logdb.Shred import Shred
from src.python.logdb.Shred import ShredTableCreator
from src.python.logdb.Shred import ShredEntrySelector
from src.python.logdb.Shred import ShredEntryClassifier
from src.python.logdb.Shred import ShredValueExtractor

class MesgShred(Shred):
    tbl_name = 'mesg'
    misfits_tbl_name = 'cont'
    create_tbl_sql = """
create table mesg(
line integer primary key references log(line),
timestamp datetime not null,
thread text not null,
logger text not null,
level text not null,
message text not null)
"""
    tbl_index_sqls = [
        'create index idx_mesg_logger on mesg(logger)',
        'create index idx_mesg_level_package on mesg(level)'
    ]
    create_misfits_sql = """
create table cont(
line integer primary key references log(line),
mesg integer key references mesg(line))    
"""
    tbl_creator = ShredTableCreator(tbl_name, create_tbl_sql, tbl_index_sqls,
        misfits_tbl_name, create_misfits_sql)

    entry_selector = ShredEntrySelector(batch_size=2000)   # selects all by default
    entry_classifier = ShredEntryClassifier()   # just one default type

    # "YYYY-MM-DD HH:MM:SS.SSS"
    # 2020-05-25 18:33:53,897 [http-nio-8080-exec-10 CapturingHttpItem] DEBUG -
    #      CRUMB request@772027798 /petclinic/oups.html		CONTEXT_SWITCH 2020-05-25 18:33:53,897 Catalina-utility-2 ==> http-nio-8080-exec-10

    extracted_val_names=['timestamp', 'thread', 'logger', 'level', 'message']
    value_extractors = {
        Shred.DEFAULT_TYPE: re.compile(
            r"^(?P<timestamp>\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2},\d{3}) \[(?P<thread>.+) (?P<logger>\S+)\] (?P<level>\S+) -( (?P<message>.*))?$")
    }
    value_extractor = ShredValueExtractor(extracted_val_names, value_extractors)

    insert_columns = ["line", "timestamp", "thread", "logger", "level", "message"]
    insert_misfits_sql = 'insert into {} (line, mesg) values (?,?)'.format(misfits_tbl_name)

    def __init__(self):
        super().__init__(
            tbl_creator=MesgShred.tbl_creator,
            entry_selector=MesgShred.entry_selector,
            entry_classifier=MesgShred.entry_classifier,
            value_extractor=MesgShred.value_extractor,
            insert_columns=MesgShred.insert_columns,
            insert_misfits_sql=MesgShred.insert_misfits_sql,
            show_misfits=False
            )

    def transform_values(self, line, entry, type, extracted_vals):
        return line, extracted_vals["timestamp"], extracted_vals["thread"], extracted_vals["logger"], extracted_vals["level"], extracted_vals["message"]

    def transform_misfits(self, line, last_good_line):
        return (line, last_good_line)

select_multiline_entries_sql = """
select line, entry
from log
where line in (
    select distinct mesg
    from cont
    UNION
    select line from cont
)
order by line
"""

if __name__ == "__main__":
    if len(sys.argv) > 1:
        logname = sys.argv[1]
    else:
        raise Exception("Missing command line parameter for log name.")

    debug_db = path.join("", logname + ".db")
    connection = createdb.connectdb(debug_db)
    mesgShred = MesgShred()
    mesgShred.initialize_tables(connection)

File: loghog/src/main/java/com/contrastsecurity/agent/loghog/shred/MesgShred.java
/* (C)2024 */
package com.contrastsecurity.agent.loghog.shred;

import static com.contrastsecurity.agent.loghog.shred.PatternGroup.*;
import static com.contrastsecurity.agent.loghog.shred.ShredEntrySelector.ALL_ENTRIES_SIGNATURE;
import static com.contrastsecurity.agent.loghog.shred.ShredEntrySelector.ALL_ENTRIES_SQL;

import com.contrastsecurity.agent.loghog.sql.SqlTableBase;
import java.util.*;
import java.util.regex.Pattern;

public class MesgShred extends Shred {

    public MesgShred() {
        super(
                new SqlTableBase(
                        MESG_TBL_NAME, MESG_TBL_CREATE_SQL, MESG_TBL_INDEX_SQLS, MESG_TBL_COLUMNS),
                new SqlTableBase(
                        CONTINUATIONS_TBL_NAME,
                        CONTINUATIONS_TBL_CREATE_SQL,
                        CONTINUATIONS_TBL_INDEX_SQLS,
                        CONTINUATIONS_TBL_COLUMNS),
                ENTRY_SELECTOR,
                ENTRY_CLASSIFIER,
                VALUE_EXTRACTOR);
    }

    @Override
    public Object[] transformValues(
            int line, String entry, String type, Map<String, Object> extractedVals) {
        return new Object[] {
            line,
            extractedVals.get("timestamp"),
            extractedVals.get("thread"),
            extractedVals.get("logger"),
            extractedVals.get("level"),
            extractedVals.get("message")
        };
    }

    @Override
    public Object[] transformMisfits(int line, int lastGoodLine) {
        return new Object[] {line, lastGoodLine};
    }

    // Shred table "mesg"
    public static final String MESG_TBL_NAME = "mesg";
    public static final String MESG_TBL_CREATE_SQL =
            """
create table mesg(
    line DECIMAL primary key references log(line),
    timestamp datetime not null,
    thread text not null,
    logger text not null,
    level text not null,
    message text not null)
""";
    public static final List<String> MESG_TBL_INDEX_SQLS =
            Arrays.asList(
                    "create index idx_mesg_line on mesg(line)",
                    "create index idx_mesg_thread on mesg(thread)");
    public static final List<String> MESG_TBL_COLUMNS =
            Arrays.asList("line", "timestamp", "thread", "logger", "level", "message");

    // Selects all lines from log table
    public static final ShredEntrySelector ENTRY_SELECTOR =
            new ShredEntrySelector(ALL_ENTRIES_SIGNATURE, ALL_ENTRIES_SQL, 2000);

    // Just one pattern type. Because we just have one pattern we don't really
    // need the classifier.  Misfit entries (continuation lines) are identified by the failure
    // of an entry to match our default pattern
    public static final ShredEntryClassifier ENTRY_CLASSIFIER = new ShredEntryClassifier();

    public static final List<String> EXTRACTED_VAL_NAMES =
            Arrays.asList(TIMESTAMP_VAR, THREAD_VAR, LOGGER_VAR, LEVEL_VAR, "message");
    public static final Map<String, Pattern> VALUE_EXTRACTORS =
            new HashMap<String, Pattern>() {
                {
                    put(
                            Shred.DEFAULT_TYPE,
                            Pattern.compile(FULL_PREAMBLE_XTRACT + "-( (?<message>.*))?$"));
                }
            };
    public static final ShredValueExtractor VALUE_EXTRACTOR =
            new ShredValueExtractor(EXTRACTED_VAL_NAMES, VALUE_EXTRACTORS);

    // Shred misfits table "cont" identifies "continuation" lines from log file that
    // don't include the normal timestamp, thread, class, log level preamble because they
    // are continuation lines from a preceding multi-line log message
    public static final String CONTINUATIONS_TBL_NAME = "cont";
    public static final String CONTINUATIONS_TBL_CREATE_SQL =
"""
create table cont(
    foreign key (line) references log(line),
    foreign key (mesg) references mesg(line)
""";
    public static final List<String> CONTINUATIONS_TBL_INDEX_SQLS = List.of();
    public static final List<String> CONTINUATIONS_TBL_COLUMNS = Arrays.asList("line", "mesg");
}
